src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.batch.sink;
+
+import com.google.common.base.Strings;
+import com.mongodb.BasicDBObjectBuilder;
+import com.mongodb.DBObject;
+import com.mongodb.hadoop.io.BSONWritable;
+import io.cdap.cdap.api.common.Bytes;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.bson.types.Decimal128;
+import org.bson.types.ObjectId;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Transforms {@link StructuredRecord} to {@link BSONWritable}.
+ */
+public class RecordToBSONWritableTransformer {
+
+  private static final String DEFAULT_ID_FIELD_NAME = "_id";
+
+  private String idFieldName;
+
+  public RecordToBSONWritableTransformer() {
+  }
+
+  /**
+   * @param idFieldName specifies which of the incoming fields should be used as an document identifier. Identifier will
+   *                be generated if no value is specified.
+   */
+  public RecordToBSONWritableTransformer(@Nullable String idFieldName) {
+    this.idFieldName = idFieldName;
+  }
+
+  /**
+   * Transforms given {@link StructuredRecord} to {@link BSONWritable}.
+   *
+   * @param original structured record to be transformed.
+   * @return {@link BSONWritable} that corresponds to the given {@link StructuredRecord}.
+   */
+  public BSONWritable transform(StructuredRecord original) {
+    StructuredRecord record = Strings.isNullOrEmpty(idFieldName) || DEFAULT_ID_FIELD_NAME.equals(idFieldName) ?
+      original : createRecordWithId(original);
+    return new BSONWritable(extractRecord(record, DEFAULT_ID_FIELD_NAME));
+  }
+
+  private StructuredRecord createRecordWithId(StructuredRecord original) {
```

Comment:
instead of this, it would be more efficient and simpler to add the id field to the transformed DBObject

Reply:
Changed. Thanks!